// @tanstack/solid-hotkeys@0.4.2/dist/createHotkey.js
import { useDefaultHotkeysOptions } from "./HotkeysProvider.js";
import { detectPlatform, formatHotkey, getHotkeyManager, rawHotkeyToParsedHotkey } from "@tanstack/hotkeys";
import { createEffect, onCleanup } from "solid-js";

//#region src/createHotkey.ts
/**
* SolidJS primitive for registering a keyboard hotkey.
*
* Uses the singleton HotkeyManager for efficient event handling.
* The callback receives both the keyboard event and a context object
* containing the hotkey string and parsed hotkey.
*
* This primitive automatically tracks reactive dependencies and updates
* the registration when options or the callback change.
*
* @param hotkey - The hotkey string (e.g., 'Mod+S', 'Escape') or RawHotkey object (supports `mod` for cross-platform)
* @param callback - The function to call when the hotkey is pressed
* @param options - Options for the hotkey behavior
*
* @example
* ```tsx
* function SaveButton() {
*   const [count, setCount] = createSignal(0)
*
*   // Callback always has access to latest count value
*   createHotkey('Mod+S', (event, { hotkey }) => {
*     console.log(`Save triggered, count is ${count()}`)
*     handleSave()
*   })
*
*   return <button onClick={() => setCount(c => c + 1)}>Count: {count()}</button>
* }
* ```
*
* @example
* ```tsx
* function Modal(props) {
*   // enabled option is reactive
*   createHotkey('Escape', () => {
*     props.onClose()
*   }, () => ({ enabled: props.isOpen }))
*
*   return <Show when={props.isOpen}>
*     <div class="modal">...</div>
*   </Show>
* }
* ```
*
* @example
* ```tsx
* function Editor() {
*   const [editorRef, setEditorRef] = createSignal<HTMLDivElement | null>(null)
*
*   // Scoped to a specific element - use accessor so hotkey waits for ref
*   createHotkey('Mod+S', save, () => ({ target: editorRef() }))
*
*   return <div ref={setEditorRef}>...</div>
* }
* ```
*/
function createHotkey(hotkey, callback, options = {}) {
	const defaultOptions = useDefaultHotkeysOptions();
	const manager = getHotkeyManager();
	let registration = null;
	createEffect(() => {
		const resolvedHotkey = typeof hotkey === "function" ? hotkey() : hotkey;
		const resolvedOptions = typeof options === "function" ? options() : options;
		const mergedOptions = {
			...defaultOptions.hotkey,
			...resolvedOptions
		};
		const platform = mergedOptions.platform ?? detectPlatform();
		const hotkeyString = typeof resolvedHotkey === "string" ? resolvedHotkey : formatHotkey(rawHotkeyToParsedHotkey(resolvedHotkey, platform));
		const resolvedTarget = "target" in mergedOptions ? mergedOptions.target ?? null : typeof document !== "undefined" ? document : null;
		if (!resolvedTarget) return;
		if (registration?.isActive) {
			registration.unregister();
			registration = null;
		}
		const { target: _target, ...optionsWithoutTarget } = mergedOptions;
		registration = manager.register(hotkeyString, callback, {
			...optionsWithoutTarget,
			target: resolvedTarget
		});
		if (registration.isActive) {
			registration.callback = callback;
			registration.setOptions(optionsWithoutTarget);
		}
		onCleanup(() => {
			if (registration?.isActive) {
				registration.unregister();
				registration = null;
			}
		});
	});
}

//#endregion
export { createHotkey };
