// @tanstack/hotkeys@0.4.2/dist/constants.js
//#region src/constants.ts
/**
* Detects the current platform based on browser navigator properties.
*
* Used internally to resolve platform-adaptive modifiers like 'Mod' (Command on Mac,
* Control elsewhere) and for platform-specific hotkey formatting.
*
* @returns The detected platform: 'mac', 'windows', or 'linux'
* @remarks Defaults to 'linux' in SSR environments where navigator is undefined
*
* @example
* ```ts
* const platform = detectPlatform() // 'mac' | 'windows' | 'linux'
* const modifier = resolveModifier('Mod', platform) // 'Meta' on Mac, 'Control' elsewhere
* ```
*/
function detectPlatform() {
	if (typeof navigator === "undefined") return "linux";
	const platform = navigator.platform?.toLowerCase() ?? "";
	const userAgent = navigator.userAgent?.toLowerCase() ?? "";
	if (platform.includes("mac") || userAgent.includes("mac")) return "mac";
	if (platform.includes("win") || userAgent.includes("win")) return "windows";
	return "linux";
}
/**
* Canonical order for modifiers in normalized hotkey strings.
*
* Defines the standard order in which modifiers should appear when formatting
* hotkeys. This ensures consistent, predictable output across the library.
*
* Order: Control → Alt → Shift → Meta
*
* @example
* ```ts
* // Input: 'Shift+Control+Meta+S'
* // Normalized: 'Control+Alt+Shift+Meta+S' (following MODIFIER_ORDER)
* ```
*/
const MODIFIER_ORDER = [
	"Control",
	"Alt",
	"Shift",
	"Meta"
];
/**
* Set of canonical modifier key names for fast lookup.
*
* Derived from `MODIFIER_ORDER` to ensure consistency. Used to detect when
* a modifier is released so we can clear non-modifier keys whose keyup events
* may have been swallowed by the OS (e.g. macOS Cmd+key combos).
*/
const MODIFIER_KEYS = new Set(MODIFIER_ORDER);
/**
* Maps modifier key aliases to their canonical form or platform-adaptive 'Mod'.
*
* This map allows users to write hotkeys using various aliases (e.g., 'Ctrl', 'Cmd', 'Option')
* which are then normalized to canonical names ('Control', 'Meta', 'Alt') or the
* platform-adaptive 'Mod' token.
*
* The 'Mod' and 'CommandOrControl' aliases are resolved at runtime via `resolveModifier()`
* based on the detected platform (Command on Mac, Control elsewhere).
*
* @remarks Case-insensitive lookups are supported via lowercase variants
*
* @example
* ```ts
* MODIFIER_ALIASES['Ctrl'] // 'Control'
* MODIFIER_ALIASES['Cmd'] // 'Meta'
* MODIFIER_ALIASES['Mod'] // 'Mod' (resolved at runtime)
* ```
*/
const MODIFIER_ALIASES = {
	Control: "Control",
	Ctrl: "Control",
	control: "Control",
	ctrl: "Control",
	Shift: "Shift",
	shift: "Shift",
	Alt: "Alt",
	Option: "Alt",
	alt: "Alt",
	option: "Alt",
	Command: "Meta",
	Cmd: "Meta",
	Meta: "Meta",
	command: "Meta",
	cmd: "Meta",
	meta: "Meta",
	CommandOrControl: "Mod",
	Mod: "Mod",
	commandorcontrol: "Mod",
	mod: "Mod"
};
/**
* Resolves the platform-adaptive 'Mod' modifier to the appropriate canonical modifier.
*
* The 'Mod' token represents the "primary modifier" on each platform:
* - macOS: 'Meta' (Command key ⌘)
* - Windows/Linux: 'Control' (Ctrl key)
*
* This enables cross-platform hotkey definitions like 'Mod+S' that automatically
* map to Command+S on Mac and Ctrl+S on Windows/Linux.
*
* @param modifier - The modifier to resolve. If 'Mod', resolves based on platform.
* @param platform - The target platform. Defaults to auto-detection.
* @returns The canonical modifier name ('Control', 'Shift', 'Alt', or 'Meta')
*
* @example
* ```ts
* resolveModifier('Mod', 'mac') // 'Meta'
* resolveModifier('Mod', 'windows') // 'Control'
* resolveModifier('Control', 'mac') // 'Control' (unchanged)
* ```
*/
function resolveModifier(modifier, platform = detectPlatform()) {
	if (modifier === "Mod") return platform === "mac" ? "Meta" : "Control";
	return modifier;
}
/**
* Set of all valid letter keys (A-Z).
*
* Used for validation and type checking. Letter keys are matched case-insensitively
* in hotkey matching, but normalized to uppercase in canonical form.
*/
const LETTER_KEYS = new Set([
	"A",
	"B",
	"C",
	"D",
	"E",
	"F",
	"G",
	"H",
	"I",
	"J",
	"K",
	"L",
	"M",
	"N",
	"O",
	"P",
	"Q",
	"R",
	"S",
	"T",
	"U",
	"V",
	"W",
	"X",
	"Y",
	"Z"
]);
/**
* Set of all valid number keys (0-9).
*
* Note: Number keys are affected by Shift (Shift+1 → '!' on US layout),
* so they're excluded from Shift-based hotkey combinations to avoid
* layout-dependent behavior.
*/
const NUMBER_KEYS = new Set([
	"0",
	"1",
	"2",
	"3",
	"4",
	"5",
	"6",
	"7",
	"8",
	"9"
]);
/**
* Set of all valid function keys (F1-F12).
*
* Function keys are commonly used for system shortcuts (e.g., F12 for DevTools,
* Alt+F4 to close windows) and application-specific commands.
*/
const FUNCTION_KEYS = new Set([
	"F1",
	"F2",
	"F3",
	"F4",
	"F5",
	"F6",
	"F7",
	"F8",
	"F9",
	"F10",
	"F11",
	"F12"
]);
/**
* Set of all valid navigation keys for cursor movement and document navigation.
*
* Includes arrow keys, Home/End (line navigation), and PageUp/PageDown (page navigation).
* These keys are commonly combined with modifiers for selection (Shift+ArrowUp) or
* navigation shortcuts (Alt+ArrowLeft for back).
*/
const NAVIGATION_KEYS = new Set([
	"ArrowUp",
	"ArrowDown",
	"ArrowLeft",
	"ArrowRight",
	"Home",
	"End",
	"PageUp",
	"PageDown"
]);
/**
* Set of all valid editing and special keys.
*
* Includes keys commonly used for text editing (Enter, Backspace, Delete, Tab) and
* special actions (Escape, Space). These keys are frequently combined with modifiers
* for editor shortcuts (Mod+Enter to submit, Shift+Tab to go back).
*/
const EDITING_KEYS = new Set([
	"Enter",
	"Escape",
	"Space",
	"Tab",
	"Backspace",
	"Delete"
]);
/**
* Set of all valid punctuation keys commonly used in keyboard shortcuts.
*
* These are the literal characters as they appear in `KeyboardEvent.key` (layout-dependent,
* typically US keyboard layout). Common shortcuts include:
* - `Mod+/` - Toggle comment
* - `Mod+[` / `Mod+]` - Indent/outdent
* - `Mod+=` / `Mod+-` - Zoom in/out
*
* Note: Punctuation keys are affected by Shift (Shift+',' → '<' on US layout),
* so they're excluded from Shift-based hotkey combinations to avoid layout-dependent behavior.
*/
const PUNCTUATION_KEYS = new Set([
	"/",
	"[",
	"]",
	"\\",
	"=",
	"-",
	",",
	".",
	";",
	"`"
]);
/**
* Maps `KeyboardEvent.code` values for punctuation keys to their canonical characters.
*
* On macOS, holding the Option (Alt) key transforms punctuation keys into special characters
* (e.g., Option+Minus → en-dash '–'), causing `event.key` to differ from the expected character.
* However, `event.code` still reports the physical key (e.g., 'Minus'). This map enables
* falling back to `event.code` for punctuation keys, similar to the existing `Key*`/`Digit*`
* fallbacks for letters and digits.
*/
const PUNCTUATION_CODE_MAP = {
	Minus: "-",
	Equal: "=",
	Slash: "/",
	BracketLeft: "[",
	BracketRight: "]",
	Backslash: "\\",
	Comma: ",",
	Period: ".",
	Backquote: "`",
	Semicolon: ";"
};
/**
* Set of all valid non-modifier keys.
*
* This is the union of all key type sets (letters, numbers, function keys, navigation,
* editing, and punctuation). Used primarily for validation to check if a key string
* is recognized and will have type-safe autocomplete support.
*
* @see {@link LETTER_KEYS}
* @see {@link NUMBER_KEYS}
* @see {@link FUNCTION_KEYS}
* @see {@link NAVIGATION_KEYS}
* @see {@link EDITING_KEYS}
* @see {@link PUNCTUATION_KEYS}
*/
const ALL_KEYS = new Set([
	...LETTER_KEYS,
	...NUMBER_KEYS,
	...FUNCTION_KEYS,
	...NAVIGATION_KEYS,
	...EDITING_KEYS,
	...PUNCTUATION_KEYS
]);
/**
* Maps key name aliases to their canonical form.
*
* Handles common variations and alternative names for keys to provide a more
* forgiving API. For example, users can write 'Esc', 'esc', or 'escape' and
* they'll all normalize to 'Escape'.
*
* This map is used internally by `normalizeKeyName()` to convert user input
* into the canonical key names used throughout the library.
*
* @remarks Case-insensitive lookups are supported via lowercase variants
*
* @example
* ```ts
* KEY_ALIASES['Esc'] // 'Escape'
* KEY_ALIASES['Del'] // 'Delete'
* KEY_ALIASES['Up'] // 'ArrowUp'
* ```
*/
const KEY_ALIASES = {
	Esc: "Escape",
	esc: "Escape",
	escape: "Escape",
	Return: "Enter",
	return: "Enter",
	enter: "Enter",
	" ": "Space",
	space: "Space",
	Spacebar: "Space",
	spacebar: "Space",
	tab: "Tab",
	backspace: "Backspace",
	Del: "Delete",
	del: "Delete",
	delete: "Delete",
	Up: "ArrowUp",
	up: "ArrowUp",
	arrowup: "ArrowUp",
	Down: "ArrowDown",
	down: "ArrowDown",
	arrowdown: "ArrowDown",
	Left: "ArrowLeft",
	left: "ArrowLeft",
	arrowleft: "ArrowLeft",
	Right: "ArrowRight",
	right: "ArrowRight",
	arrowright: "ArrowRight",
	home: "Home",
	end: "End",
	pageup: "PageUp",
	pagedown: "PageDown",
	PgUp: "PageUp",
	PgDn: "PageDown",
	pgup: "PageUp",
	pgdn: "PageDown"
};
/**
* Normalizes a key name to its canonical form.
*
* Converts various key name formats (aliases, case variations) into the standard
* canonical names used throughout the library. This enables a more forgiving API
* where users can write keys in different ways and still get correct behavior.
*
* Normalization rules:
* 1. Check aliases first (e.g., 'Esc' → 'Escape', 'Del' → 'Delete')
* 2. Single letters → uppercase (e.g., 'a' → 'A', 's' → 'S')
* 3. Function keys → uppercase (e.g., 'f1' → 'F1', 'F12' → 'F12')
* 4. Other keys → returned as-is (already canonical or unknown)
*
* @param key - The key name to normalize (can be an alias, lowercase, etc.)
* @returns The canonical key name
*
* @example
* ```ts
* normalizeKeyName('esc') // 'Escape'
* normalizeKeyName('a') // 'A'
* normalizeKeyName('f1') // 'F1'
* normalizeKeyName('ArrowUp') // 'ArrowUp' (already canonical)
* ```
*/
function isSingleLetterKey(key) {
	return /^\p{Letter}$/u.test(key);
}
function normalizeKeyName(key) {
	if (key in KEY_ALIASES) return KEY_ALIASES[key];
	if (isSingleLetterKey(key)) {
		const upper = key.toUpperCase();
		return upper.length === 1 ? upper : key;
	}
	const upperKey = key.toUpperCase();
	if (/^F([1-9]|1[0-2])$/.test(upperKey)) return upperKey;
	return key;
}
/**
* Modifier key symbols for macOS display.
*
* Used by formatting functions to display hotkeys with macOS-style symbols
* (e.g., ⌘ for Command, ⌃ for Control) instead of text labels. This provides
* a native macOS look and feel in hotkey displays.
*
* @example
* ```ts
* MAC_MODIFIER_SYMBOLS['Meta'] // '⌘'
* MAC_MODIFIER_SYMBOLS['Control'] // '⌃'
* MAC_MODIFIER_SYMBOLS['Alt'] // '⌥'
* MAC_MODIFIER_SYMBOLS['Shift'] // '⇧'
* ```
*/
const MAC_MODIFIER_SYMBOLS = {
	Control: "⌃",
	Alt: "⌥",
	Shift: "⇧",
	Meta: "⌘"
};
/**
* Modifier key labels for Windows/Linux display.
*
* Used by formatting functions to display hotkeys with standard text labels
* (e.g., 'Ctrl' for Control, 'Win' for Meta/Windows key) instead of symbols.
* This provides a familiar Windows/Linux look and feel in hotkey displays.
*
* @example
* ```ts
* STANDARD_MODIFIER_LABELS['Control'] // 'Ctrl'
* STANDARD_MODIFIER_LABELS['Meta'] // 'Win'
* STANDARD_MODIFIER_LABELS['Alt'] // 'Alt'
* STANDARD_MODIFIER_LABELS['Shift'] // 'Shift'
* ```
*/
const STANDARD_MODIFIER_LABELS = {
	Control: "Ctrl",
	Alt: "Alt",
	Shift: "Shift",
	Meta: "Win"
};
/**
* Special key symbols for display formatting.
*
* Maps certain keys to their visual symbols for better readability in hotkey displays.
* Used by formatting functions to show symbols like ↑ for ArrowUp or ↵ for Enter
* instead of text labels.
*
* @example
* ```ts
* KEY_DISPLAY_SYMBOLS['ArrowUp'] // '↑'
* KEY_DISPLAY_SYMBOLS['Enter'] // '↵'
* KEY_DISPLAY_SYMBOLS['Escape'] // 'Esc'
* KEY_DISPLAY_SYMBOLS['Space'] // '␣'
* ```
*/
const KEY_DISPLAY_SYMBOLS = {
	ArrowUp: "↑",
	ArrowDown: "↓",
	ArrowLeft: "←",
	ArrowRight: "→",
	Enter: "↵",
	Escape: "Esc",
	Backspace: "⌫",
	Delete: "⌦",
	Tab: "⇥",
	Space: "␣"
};

//#endregion
export { ALL_KEYS, EDITING_KEYS, FUNCTION_KEYS, KEY_DISPLAY_SYMBOLS, LETTER_KEYS, MAC_MODIFIER_SYMBOLS, MODIFIER_ALIASES, MODIFIER_KEYS, MODIFIER_ORDER, NAVIGATION_KEYS, NUMBER_KEYS, PUNCTUATION_CODE_MAP, PUNCTUATION_KEYS, STANDARD_MODIFIER_LABELS, detectPlatform, isSingleLetterKey, normalizeKeyName, resolveModifier };
