// @tanstack/hotkeys@0.4.2/dist/format.js
import { KEY_DISPLAY_SYMBOLS, MAC_MODIFIER_SYMBOLS, MODIFIER_ORDER, STANDARD_MODIFIER_LABELS, detectPlatform } from "./constants.js";
import { parseHotkey } from "./parse.js";

//#region src/format.ts
/**
* Converts a hotkey sequence array to a display string.
*
* @param sequence - Array of hotkey strings that form the sequence
* @returns A space-separated string (e.g. ['G','G'] → 'G G')
*
* @example
* ```ts
* formatHotkeySequence(['G', 'G'])      // 'G G'
* formatHotkeySequence(['D', 'I', 'W']) // 'D I W'
* ```
*/
function formatHotkeySequence(sequence) {
	return sequence.join(" ");
}
/**
* Converts a ParsedHotkey back to a hotkey string.
*
* @param parsed - The parsed hotkey object
* @returns A hotkey string in canonical form
*
* @example
* ```ts
* formatHotkey({ key: 'S', ctrl: true, shift: true, alt: false, meta: false, modifiers: ['Control', 'Shift'] })
* // Returns: 'Control+Shift+S'
* ```
*/
function formatHotkey(parsed) {
	const parts = [];
	for (const modifier of MODIFIER_ORDER) if (parsed.modifiers.includes(modifier)) parts.push(modifier);
	parts.push(parsed.key);
	return parts.join("+");
}
/**
* Formats a hotkey for display in a user interface.
*
* On macOS, uses symbols (⌘⇧S).
* On Windows/Linux, uses text (Ctrl+Shift+S).
*
* @param hotkey - The hotkey string or ParsedHotkey to format
* @param options - Formatting options
* @returns A formatted string suitable for display
*
* @example
* ```ts
* formatForDisplay('Mod+Shift+S', { platform: 'mac' })
* // Returns: '⇧⌘S'
*
* formatForDisplay('Mod+Shift+S', { platform: 'windows' })
* // Returns: 'Ctrl+Shift+S'
*
* formatForDisplay('Escape')
* // Returns: 'Esc' (on all platforms)
* ```
*/
function formatForDisplay(hotkey, options = {}) {
	const platform = options.platform ?? detectPlatform();
	const parsed = typeof hotkey === "string" ? parseHotkey(hotkey, platform) : hotkey;
	if (platform === "mac") return formatForMac(parsed);
	return formatForStandard(parsed);
}
/**
* Formats a hotkey for macOS display using symbols.
*/
function formatForMac(parsed) {
	const parts = [];
	for (const modifier of MODIFIER_ORDER) if (parsed.modifiers.includes(modifier)) parts.push(MAC_MODIFIER_SYMBOLS[modifier]);
	const keyDisplay = KEY_DISPLAY_SYMBOLS[parsed.key] ?? parsed.key;
	parts.push(keyDisplay);
	return parts.join("");
}
/**
* Formats a hotkey for Windows/Linux display using text labels.
*/
function formatForStandard(parsed) {
	const parts = [];
	for (const modifier of MODIFIER_ORDER) if (parsed.modifiers.includes(modifier)) parts.push(STANDARD_MODIFIER_LABELS[modifier]);
	const keyDisplay = KEY_DISPLAY_SYMBOLS[parsed.key] ?? parsed.key;
	parts.push(keyDisplay);
	return parts.join("+");
}
/**
* Formats a hotkey using platform-agnostic labels.
* Uses 'Cmd' on Mac and 'Ctrl' for Control, etc.
*
* @param hotkey - The hotkey string or ParsedHotkey to format
* @param platform - The target platform
* @returns A formatted string with platform-appropriate labels
*/
function formatWithLabels(hotkey, platform = detectPlatform()) {
	const parsed = typeof hotkey === "string" ? parseHotkey(hotkey, platform) : hotkey;
	const parts = [];
	const labels = {
		Control: "Ctrl",
		Alt: platform === "mac" ? "Option" : "Alt",
		Shift: "Shift",
		Meta: platform === "mac" ? "Cmd" : "Win"
	};
	for (const modifier of MODIFIER_ORDER) if (parsed.modifiers.includes(modifier)) parts.push(labels[modifier] ?? modifier);
	parts.push(parsed.key);
	return parts.join("+");
}
/**
* Maps canonical modifier names to debugging-friendly labels per platform.
*/
const MODIFIER_DEBUG_LABELS = {
	mac: {
		Meta: "Mod (Cmd)",
		Control: "Ctrl",
		Alt: "Opt",
		Shift: "Shift"
	},
	windows: {
		Control: "Mod (Ctrl)",
		Meta: "Win",
		Alt: "Alt",
		Shift: "Shift"
	},
	linux: {
		Control: "Mod (Ctrl)",
		Meta: "Super",
		Alt: "Alt",
		Shift: "Shift"
	}
};
/**
* Formats a single key name for debugging/devtools display.
*
* Unlike `formatForDisplay` which formats full hotkey strings for end-user UIs,
* this function formats individual key names (from `event.key`) with rich
* platform-aware labels suitable for debugging tools and developer-facing displays.
*
* Features:
* - Modifier keys show their platform role (e.g., "Mod (Cmd)" for Meta on Mac)
* - On macOS, modifier keys are prefixed with their symbol (e.g., "⌘ Mod (Cmd)")
* - Special keys use display symbols (ArrowUp -> "↑", Escape -> "Esc")
* - Regular keys pass through unchanged
*
* @param key - A single key name (e.g., "Meta", "Shift", "ArrowUp", "A")
* @param options - Formatting options
* @returns A formatted label suitable for debugging display
*
* @example
* ```ts
* // On macOS:
* formatKeyForDebuggingDisplay('Meta')    // '⌘ Mod (Cmd)'
* formatKeyForDebuggingDisplay('Control') // '⌃ Ctrl'
* formatKeyForDebuggingDisplay('Alt')     // '⌥ Opt'
* formatKeyForDebuggingDisplay('Shift')   // '⇧ Shift'
*
* // On Windows:
* formatKeyForDebuggingDisplay('Control') // 'Mod (Ctrl)'
* formatKeyForDebuggingDisplay('Meta')    // 'Win'
*
* // Special keys (all platforms):
* formatKeyForDebuggingDisplay('ArrowUp') // '↑'
* formatKeyForDebuggingDisplay('Escape')  // 'Esc'
* formatKeyForDebuggingDisplay('Space')   // '␣'
*
* // Regular keys pass through:
* formatKeyForDebuggingDisplay('A')       // 'A'
*
* // With source: 'code', values pass through unchanged:
* formatKeyForDebuggingDisplay('MetaLeft', { source: 'code' })  // 'MetaLeft'
* formatKeyForDebuggingDisplay('KeyA', { source: 'code' })      // 'KeyA'
* ```
*/
function formatKeyForDebuggingDisplay(key, options = {}) {
	if (options.source === "code") return key;
	const platform = options.platform ?? detectPlatform();
	const modLabel = MODIFIER_DEBUG_LABELS[platform]?.[key];
	if (modLabel) {
		if (platform === "mac") {
			const symbol = MAC_MODIFIER_SYMBOLS[key];
			if (symbol) return `${symbol} ${modLabel}`;
		}
		return modLabel;
	}
	const symbol = KEY_DISPLAY_SYMBOLS[key];
	if (symbol) return symbol;
	return key;
}

//#endregion
export { formatForDisplay, formatHotkey, formatHotkeySequence, formatKeyForDebuggingDisplay, formatWithLabels };
