// @tanstack/hotkeys@0.4.2/dist/manager.utils.js
//#region src/manager.utils.ts
/**
* Default options for hotkey/sequence registration.
* Omitted: platform, target (resolved at registration), requireReset (HotkeyManager only).
*/
const defaultHotkeyOptions = {
	preventDefault: true,
	stopPropagation: true,
	eventType: "keydown",
	enabled: true,
	ignoreInputs: true,
	conflictBehavior: "warn"
};
/**
* Computes the default ignoreInputs value based on the hotkey.
* Ctrl/Meta shortcuts and Escape fire in inputs; single keys and Shift/Alt combos are ignored.
*/
function getDefaultIgnoreInputs(parsedHotkey) {
	if (parsedHotkey.ctrl || parsedHotkey.meta) return false;
	if (parsedHotkey.key === "Escape") return false;
	return true;
}
/**
* Checks if an element is an input-like element that should be ignored for hotkeys.
*
* This includes:
* - HTMLInputElement (all input types except button, submit, reset)
* - HTMLTextAreaElement
* - HTMLSelectElement
* - Elements with contentEditable enabled
*
* Button-type inputs (button, submit, reset) are excluded so hotkeys like
* Mod+S and Escape fire when the user has tabbed to a form button.
*/
function isInputElement(element) {
	if (!element) return false;
	if (element instanceof HTMLInputElement) {
		const type = element.type.toLowerCase();
		if (type === "button" || type === "submit" || type === "reset") return false;
		return true;
	}
	if (element instanceof HTMLTextAreaElement || element instanceof HTMLSelectElement) return true;
	if (element instanceof HTMLElement && element.isContentEditable) return true;
	return false;
}
/**
* Checks if an event is for the given target (originated from or bubbled to it).
*
* For document/window targets, also accepts document.documentElement as currentTarget
* to handle Brave and other browsers where currentTarget may be documentElement
* instead of document when listeners are attached to document.
*/
function isEventForTarget(event, target) {
	if (target === document || target === window) return event.currentTarget === target || event.currentTarget === document.documentElement;
	if (target === window) return event.currentTarget === window || event.currentTarget === document || event.currentTarget === document.documentElement;
	if (target instanceof HTMLElement) {
		if (event.currentTarget === target) return true;
		if (event.target instanceof Node && target.contains(event.target)) return true;
	}
	return false;
}
/**
* Handles conflicts between registrations based on conflict behavior.
*
* @param conflictingId - The ID of the conflicting registration
* @param keyDisplay - Display string for the conflicting key/sequence (for error messages)
* @param conflictBehavior - How to handle the conflict
* @param unregister - Function to unregister by ID
*/
function handleConflict(conflictingId, keyDisplay, conflictBehavior, unregister) {
	if (conflictBehavior === "allow") return;
	if (conflictBehavior === "warn") {
		console.warn(`'${keyDisplay}' is already registered. Multiple handlers will be triggered. Use conflictBehavior: 'replace' to replace the existing handler, or conflictBehavior: 'allow' to suppress this warning.`);
		return;
	}
	if (conflictBehavior === "error") throw new Error(`'${keyDisplay}' is already registered. Use conflictBehavior: 'replace' to replace the existing handler, or conflictBehavior: 'allow' to allow multiple registrations.`);
	unregister(conflictingId);
}

//#endregion
export { defaultHotkeyOptions, getDefaultIgnoreInputs, handleConflict, isEventForTarget, isInputElement };
