// @tanstack/hotkeys@0.4.2/dist/match.js
import { PUNCTUATION_CODE_MAP, detectPlatform, isSingleLetterKey, normalizeKeyName } from "./constants.js";
import { parseHotkey } from "./parse.js";

//#region src/match.ts
/**
* Checks if a KeyboardEvent matches a hotkey.
*
* Uses the `key` property from KeyboardEvent for matching, with a fallback to `code`
* for letter keys, digit keys (0-9), and punctuation keys when `key` produces special
* characters (e.g., macOS Option+letter, Shift+number, or Option+punctuation).
* Letter keys are matched case-insensitively.
*
* Also handles "dead key" events where `event.key` is `'Dead'` instead of the expected
* character. This commonly occurs on macOS with Option+letter combinations (e.g., Option+E,
* Option+I, Option+U, Option+N) and on Windows/Linux with international keyboard layouts.
* In these cases, `event.code` is used to determine the physical key.
*
* @param event - The KeyboardEvent to check
* @param hotkey - The hotkey string or ParsedHotkey to match against
* @param platform - The target platform for resolving 'Mod' (defaults to auto-detection)
* @returns True if the event matches the hotkey
*
* @example
* ```ts
* document.addEventListener('keydown', (event) => {
*   if (matchesKeyboardEvent(event, 'Mod+S')) {
*     event.preventDefault()
*     handleSave()
*   }
* })
* ```
*/
function matchesKeyboardEvent(event, hotkey, platform = detectPlatform()) {
	const parsed = typeof hotkey === "string" ? parseHotkey(hotkey, platform) : hotkey;
	if (event.ctrlKey !== parsed.ctrl) return false;
	if (event.shiftKey !== parsed.shift) return false;
	if (event.altKey !== parsed.alt) return false;
	if (event.metaKey !== parsed.meta) return false;
	const eventKey = normalizeKeyName(event.key);
	const hotkeyKey = parsed.key;
	if (eventKey !== "Dead" && eventKey.length === 1 && hotkeyKey.length === 1) {
		if (eventKey.toUpperCase() === hotkeyKey.toUpperCase()) return true;
		if (isSingleLetterKey(eventKey) && (/^[A-Za-z]$/.test(eventKey) || !event.altKey)) return false;
	}
	if (event.code && (eventKey === "Dead" || eventKey.length === 1 && hotkeyKey.length === 1)) {
		if (event.code.startsWith("Key")) {
			const codeLetter = event.code.slice(3);
			if (codeLetter.length === 1 && /^[A-Za-z]$/.test(codeLetter)) return codeLetter.toUpperCase() === hotkeyKey.toUpperCase();
		}
		if (event.code.startsWith("Digit")) {
			const codeDigit = event.code.slice(5);
			if (codeDigit.length === 1 && /^[0-9]$/.test(codeDigit)) return codeDigit === hotkeyKey;
		}
		if (event.code in PUNCTUATION_CODE_MAP) return PUNCTUATION_CODE_MAP[event.code] === hotkeyKey;
		return false;
	}
	return eventKey === hotkeyKey;
}
/**
* Creates a keyboard event handler that calls the callback when the hotkey matches.
*
* @param hotkey - The hotkey string or ParsedHotkey to match
* @param callback - The function to call when the hotkey matches
* @param options - Options for matching and handling
* @returns A function that can be used as an event handler
*
* @example
* ```ts
* const handler = createHotkeyHandler('Mod+S', (event, { hotkey, parsedHotkey }) => {
*   console.log(`${hotkey} was pressed`)
*   handleSave()
* })
*
* document.addEventListener('keydown', handler)
* ```
*/
function createHotkeyHandler(hotkey, callback, options = {}) {
	const { preventDefault = true, stopPropagation = true, platform } = options;
	const resolvedPlatform = platform ?? detectPlatform();
	const hotkeyString = typeof hotkey === "string" ? hotkey : formatParsedHotkey(hotkey);
	const parsed = typeof hotkey === "string" ? parseHotkey(hotkey, resolvedPlatform) : hotkey;
	const context = {
		hotkey: hotkeyString,
		parsedHotkey: parsed
	};
	return (event) => {
		if (matchesKeyboardEvent(event, parsed, resolvedPlatform)) {
			if (preventDefault) event.preventDefault();
			if (stopPropagation) event.stopPropagation();
			callback(event, context);
		}
	};
}
/**
* Creates a handler that matches multiple hotkeys.
*
* @param handlers - A map of hotkey strings to their handlers
* @param options - Options for matching and handling
* @returns A function that can be used as an event handler
*
* @example
* ```ts
* const handler = createMultiHotkeyHandler({
*   'Mod+S': (event, { hotkey }) => handleSave(),
*   'Mod+Z': (event, { hotkey }) => handleUndo(),
*   'Mod+Shift+Z': (event, { hotkey }) => handleRedo(),
* })
*
* document.addEventListener('keydown', handler)
* ```
*/
function createMultiHotkeyHandler(handlers, options = {}) {
	const { preventDefault = true, stopPropagation = true, platform } = options;
	const resolvedPlatform = platform ?? detectPlatform();
	const parsedHandlers = Object.entries(handlers).filter((entry) => Boolean(entry[1])).map(([hotkey, handler]) => {
		const parsed = parseHotkey(hotkey, resolvedPlatform);
		return {
			parsed,
			handler,
			context: {
				hotkey,
				parsedHotkey: parsed
			}
		};
	});
	return (event) => {
		for (const { parsed, handler, context } of parsedHandlers) if (matchesKeyboardEvent(event, parsed, resolvedPlatform)) {
			if (preventDefault) event.preventDefault();
			if (stopPropagation) event.stopPropagation();
			handler(event, context);
			return;
		}
	};
}
/**
* Formats a ParsedHotkey back to a hotkey string.
* Used internally to provide the hotkey string in callback context.
*/
function formatParsedHotkey(parsed) {
	const parts = [];
	if (parsed.ctrl) parts.push("Control");
	if (parsed.alt) parts.push("Alt");
	if (parsed.shift) parts.push("Shift");
	if (parsed.meta) parts.push("Meta");
	parts.push(parsed.key);
	return parts.join("+");
}

//#endregion
export { createHotkeyHandler, createMultiHotkeyHandler, matchesKeyboardEvent };
