// @tanstack/hotkeys@0.4.2/dist/parse.js
import { MODIFIER_ALIASES, MODIFIER_ORDER, detectPlatform, normalizeKeyName, resolveModifier } from "./constants.js";

//#region src/parse.ts
/**
* Parses a hotkey string into its component parts.
*
* @param hotkey - The hotkey string to parse (e.g., 'Mod+Shift+S')
* @param platform - The target platform for resolving 'Mod' (defaults to auto-detection)
* @returns A ParsedHotkey object with the key and modifier flags
*
* @example
* ```ts
* parseHotkey('Mod+S') // On Mac: { key: 'S', ctrl: false, shift: false, alt: false, meta: true, modifiers: ['Meta'] }
* parseHotkey('Mod+S') // On Windows: { key: 'S', ctrl: true, shift: false, alt: false, meta: false, modifiers: ['Control'] }
* parseHotkey('Control+Shift+A') // { key: 'A', ctrl: true, shift: true, alt: false, meta: false, modifiers: ['Control', 'Shift'] }
* ```
*/
function parseHotkey(hotkey, platform = detectPlatform()) {
	const parts = hotkey.split("+");
	const modifiers = /* @__PURE__ */ new Set();
	let key = "";
	for (let i = 0; i < parts.length; i++) {
		const part = parts[i].trim();
		if (i === parts.length - 1) key = normalizeKeyName(part);
		else {
			const alias = MODIFIER_ALIASES[part] ?? MODIFIER_ALIASES[part.toLowerCase()];
			if (alias) {
				const resolved = resolveModifier(alias, platform);
				modifiers.add(resolved);
			} else if (parts.length === 1) key = normalizeKeyName(part);
		}
	}
	if (!key && parts.length > 0) key = normalizeKeyName(parts[parts.length - 1].trim());
	return {
		key,
		ctrl: modifiers.has("Control"),
		shift: modifiers.has("Shift"),
		alt: modifiers.has("Alt"),
		meta: modifiers.has("Meta"),
		modifiers: MODIFIER_ORDER.filter((m) => modifiers.has(m))
	};
}
/**
* Converts a RawHotkey object to a ParsedHotkey.
* Optional modifier booleans default to false; modifiers array is derived from them.
* When `mod` is true, it is resolved to Control or Meta based on platform.
*
* @param raw - The raw hotkey object
* @param platform - The target platform for resolving 'Mod' (defaults to auto-detection)
* @returns A ParsedHotkey suitable for matching and formatting
*
* @example
* ```ts
* rawHotkeyToParsedHotkey({ key: 'Escape' })
* // { key: 'Escape', ctrl: false, shift: false, alt: false, meta: false, modifiers: [] }
*
* rawHotkeyToParsedHotkey({ key: 'S', mod: true }, 'mac')
* // { key: 'S', ctrl: false, shift: false, alt: false, meta: true, modifiers: ['Meta'] }
*
* rawHotkeyToParsedHotkey({ key: 'S', mod: true, shift: true }, 'windows')
* // { key: 'S', ctrl: true, shift: true, alt: false, meta: false, modifiers: ['Control', 'Shift'] }
* ```
*/
function rawHotkeyToParsedHotkey(raw, platform = detectPlatform()) {
	let ctrl = raw.ctrl ?? false;
	const shift = raw.shift ?? false;
	const alt = raw.alt ?? false;
	let meta = raw.meta ?? false;
	if (raw.mod) if (resolveModifier("Mod", platform) === "Control") ctrl = true;
	else meta = true;
	const modifiers = MODIFIER_ORDER.filter((m) => {
		switch (m) {
			case "Control": return ctrl;
			case "Shift": return shift;
			case "Alt": return alt;
			case "Meta": return meta;
			default: return false;
		}
	});
	return {
		key: raw.key,
		ctrl,
		shift,
		alt,
		meta,
		modifiers
	};
}
/**
* Normalizes a hotkey string to its canonical form.
*
* The canonical form uses:
* - Full modifier names (Control, Alt, Shift, Meta)
* - Modifiers in order: Control+Alt+Shift+Meta
* - Uppercase letters for single-character keys
* - Proper casing for special keys (Escape, not escape)
*
* @param hotkey - The hotkey string to normalize
* @param platform - The target platform for resolving 'Mod' (defaults to auto-detection)
* @returns The normalized hotkey string
*
* @example
* ```ts
* normalizeHotkey('mod+shift+s') // On Mac: 'Shift+Meta+S'
* normalizeHotkey('ctrl+a') // 'Control+A'
* normalizeHotkey('esc') // 'Escape'
* ```
*/
function normalizeHotkey(hotkey, platform = detectPlatform()) {
	const parsed = parseHotkey(hotkey, platform);
	const parts = [];
	for (const modifier of MODIFIER_ORDER) if (parsed.modifiers.includes(modifier)) parts.push(modifier);
	parts.push(parsed.key);
	return parts.join("+");
}
/**
* Checks if a string represents a modifier key.
*
* @param key - The string to check
* @returns True if the string is a recognized modifier
*/
function isModifier(key) {
	return key in MODIFIER_ALIASES || key.toLowerCase() in MODIFIER_ALIASES;
}
/**
* Parses a KeyboardEvent into a ParsedHotkey object.
*
* This function extracts the key and modifier state from a keyboard event
* and converts it into the same format used by `parseHotkey()`.
*
* @param event - The KeyboardEvent to parse
* @param platform - The target platform for resolving modifiers (defaults to auto-detection)
* @returns A ParsedHotkey object representing the keyboard event
*
* @example
* ```ts
* document.addEventListener('keydown', (event) => {
*   const parsed = parseKeyboardEvent(event)
*   console.log(parsed) // { key: 'S', ctrl: true, shift: false, ... }
* })
* ```
*/
function parseKeyboardEvent(event) {
	const normalizedKey = normalizeKeyName(event.key);
	const modifiers = [];
	if (event.ctrlKey) modifiers.push("Control");
	if (event.altKey) modifiers.push("Alt");
	if (event.shiftKey) modifiers.push("Shift");
	if (event.metaKey) modifiers.push("Meta");
	return {
		key: normalizedKey,
		ctrl: event.ctrlKey,
		shift: event.shiftKey,
		alt: event.altKey,
		meta: event.metaKey,
		modifiers
	};
}
/**
* Converts a KeyboardEvent directly to a hotkey string.
*
* This is a convenience function that combines `parseKeyboardEvent()` and formatting.
* The resulting hotkey string uses canonical modifier names (Control, Alt, Shift, Meta)
* and is suitable for use with `useHotkey()` and other hotkey functions.
*
* @param event - The KeyboardEvent to convert
* @param platform - The target platform (defaults to auto-detection)
* @returns A hotkey string in canonical form (e.g., 'Control+Shift+S')
*
* @example
* ```ts
* document.addEventListener('keydown', (event) => {
*   const hotkey = keyboardEventToHotkey(event)
*   console.log(hotkey) // 'Control+Shift+S'
*   useHotkey(hotkey, () => console.log('Shortcut triggered'))
* })
* ```
*/
function keyboardEventToHotkey(event) {
	const parsed = parseKeyboardEvent(event);
	const parts = [];
	for (const modifier of MODIFIER_ORDER) if (parsed.modifiers.includes(modifier)) parts.push(modifier);
	parts.push(parsed.key);
	return parts.join("+");
}
/**
* Checks if a KeyboardEvent represents a modifier-only key press.
*
* Modifier-only keys are keys like 'Control', 'Shift', 'Alt', 'Meta', etc.
* that don't have an associated character or action key. This is useful
* for filtering out modifier key presses when recording shortcuts.
*
* @param event - The KeyboardEvent to check
* @returns True if the event represents a modifier-only key
*
* @example
* ```ts
* document.addEventListener('keydown', (event) => {
*   if (isModifierKey(event)) {
*     console.log('Modifier key pressed, waiting for action key...')
*     return
*   }
*   // Process non-modifier key
* })
* ```
*/
function isModifierKey(event) {
	const key = event.key;
	return key === "Control" || key === "Shift" || key === "Alt" || key === "Meta" || key === "Command" || key === "OS" || key === "Win";
}
/**
* Checks if a hotkey or ParsedHotkey contains at least one non-modifier key.
*
* This is useful for validating that a recorded hotkey is complete and not
* just a combination of modifiers without an action key.
*
* @param hotkey - The hotkey string or ParsedHotkey to check
* @param platform - The target platform for parsing (defaults to auto-detection)
* @returns True if the hotkey contains at least one non-modifier key
*
* @example
* ```ts
* hasNonModifierKey('Control+Shift+S') // true
* hasNonModifierKey('Control+Shift') // false (no action key)
* hasNonModifierKey(parseHotkey('Mod+A')) // true
* ```
*/
function hasNonModifierKey(hotkey, platform = detectPlatform()) {
	const parsed = typeof hotkey === "string" ? parseHotkey(hotkey, platform) : hotkey;
	return !isModifier(parsed.key) && parsed.key.length > 0;
}
/**
* Converts a hotkey string to use 'Mod' format for portability.
*
* On macOS, converts 'Meta' to 'Mod'. On Windows/Linux, converts 'Control' to 'Mod'.
* This enables cross-platform hotkey definitions that work consistently.
*
* @param hotkey - The hotkey string to convert
* @param platform - The target platform (defaults to auto-detection)
* @returns The hotkey string with 'Mod' format applied
*
* @example
* ```ts
* convertToModFormat('Meta+S', 'mac') // 'Mod+S'
* convertToModFormat('Control+S', 'windows') // 'Mod+S'
* convertToModFormat('Control+Meta+S', 'mac') // 'Control+Meta+S' (both present, no conversion)
* ```
*/
function convertToModFormat(hotkey, platform = detectPlatform()) {
	const parsed = parseHotkey(hotkey, platform);
	if (platform === "mac" && parsed.meta && !parsed.ctrl) return hotkey.split("+").map((part) => part === "Meta" ? "Mod" : part).join("+");
	else if (platform !== "mac" && parsed.ctrl && !parsed.meta) return hotkey.split("+").map((part) => part === "Control" ? "Mod" : part).join("+");
	return hotkey;
}

//#endregion
export { convertToModFormat, hasNonModifierKey, isModifier, isModifierKey, keyboardEventToHotkey, normalizeHotkey, parseHotkey, parseKeyboardEvent, rawHotkeyToParsedHotkey };
