// @tanstack/hotkeys@0.4.2/dist/sequence-manager.js
import { detectPlatform } from "./constants.js";
import { parseHotkey } from "./parse.js";
import { formatHotkeySequence } from "./format.js";
import { matchesKeyboardEvent } from "./match.js";
import { defaultHotkeyOptions, getDefaultIgnoreInputs, handleConflict, isEventForTarget, isInputElement } from "./manager.utils.js";
import { Store } from "@tanstack/store";

//#region src/sequence-manager.ts
/**
* Default timeout between keys in a sequence (in milliseconds).
*/
const DEFAULT_SEQUENCE_TIMEOUT = 1e3;
let sequenceIdCounter = 0;
/**
* Generates a unique ID for sequence registrations.
*/
function generateSequenceId() {
	return `sequence_${++sequenceIdCounter}`;
}
/**
* Returns a canonical string for sequence conflict comparison.
*/
function sequenceKey(sequence) {
	return sequence.join("|");
}
/**
* Manages keyboard sequence matching for Vim-style shortcuts.
*
* This class allows registering multi-key sequences like 'g g' or 'd d'
* that trigger callbacks when the full sequence is pressed within
* a configurable timeout.
*
* @example
* ```ts
* const matcher = SequenceManager.getInstance()
*
* // Register 'g g' to go to top
* const unregister = matcher.register(['G', 'G'], (event, context) => {
*   scrollToTop()
* }, { timeout: 500 })
*
* // Later, to unregister:
* unregister()
* ```
*/
/**
* Builds a devtools view from an internal registration.
*/
function toRegistrationView(reg) {
	return {
		id: reg.id,
		sequence: reg.sequence,
		options: reg.options,
		target: reg.target,
		triggerCount: reg.triggerCount
	};
}
var SequenceManager = class SequenceManager {
	static #instance = null;
	#registrations = /* @__PURE__ */ new Map();
	#targetListeners = /* @__PURE__ */ new Map();
	#targetRegistrations = /* @__PURE__ */ new Map();
	#platform;
	constructor() {
		this.registrations = new Store(/* @__PURE__ */ new Map());
		this.#platform = detectPlatform();
	}
	/**
	* Gets the singleton instance of SequenceManager.
	*/
	static getInstance() {
		if (!SequenceManager.#instance) SequenceManager.#instance = new SequenceManager();
		return SequenceManager.#instance;
	}
	/**
	* Resets the singleton instance. Useful for testing.
	*/
	static resetInstance() {
		if (SequenceManager.#instance) {
			SequenceManager.#instance.destroy();
			SequenceManager.#instance = null;
		}
	}
	/**
	* Registers a hotkey sequence handler.
	*
	* @param sequence - Array of hotkey strings that form the sequence
	* @param callback - Function to call when the sequence is completed
	* @param options - Options for the sequence behavior
	* @returns A handle to update or unregister the sequence
	*/
	register(sequence, callback, options = {}) {
		if (sequence.length === 0) throw new Error("Sequence must contain at least one hotkey");
		const id = generateSequenceId();
		const platform = options.platform ?? this.#platform;
		const parsedSequence = sequence.map((hotkey) => parseHotkey(hotkey, platform));
		const target = options.target ?? (typeof document !== "undefined" ? document : {});
		const conflictBehavior = options.conflictBehavior ?? "warn";
		const conflictingRegistration = this.#findConflictingSequence(sequence, target);
		if (conflictingRegistration) handleConflict(conflictingRegistration.id, formatHotkeySequence(sequence), conflictBehavior, (regId) => this.#unregister(regId));
		const firstStep = parsedSequence[0];
		const resolvedIgnoreInputs = options.ignoreInputs ?? getDefaultIgnoreInputs(firstStep);
		const registration = {
			id,
			sequence,
			parsedSequence,
			callback,
			options: {
				...defaultHotkeyOptions,
				timeout: DEFAULT_SEQUENCE_TIMEOUT,
				...options,
				platform,
				ignoreInputs: resolvedIgnoreInputs
			},
			target,
			currentIndex: 0,
			lastKeyTime: 0,
			triggerCount: 0
		};
		this.#registrations.set(id, registration);
		this.registrations.setState((prev) => new Map(prev).set(id, toRegistrationView(registration)));
		if (!this.#targetRegistrations.has(target)) this.#targetRegistrations.set(target, /* @__PURE__ */ new Set());
		this.#targetRegistrations.get(target).add(id);
		this.#ensureListenersForTarget(target);
		const manager = this;
		return {
			get id() {
				return id;
			},
			get isActive() {
				return manager.#registrations.has(id);
			},
			get callback() {
				return manager.#registrations.get(id)?.callback ?? callback;
			},
			set callback(newCallback) {
				const reg = manager.#registrations.get(id);
				if (reg) reg.callback = newCallback;
			},
			setOptions: (newOptions) => {
				const reg = manager.#registrations.get(id);
				if (reg) {
					reg.options = {
						...reg.options,
						...newOptions
					};
					manager.registrations.setState((prev) => new Map(prev).set(id, toRegistrationView(reg)));
				}
			},
			unregister: () => {
				manager.#unregister(id);
			}
		};
	}
	/**
	* Unregisters a sequence by its registration ID.
	*/
	#unregister(id) {
		const registration = this.#registrations.get(id);
		if (!registration) return;
		const target = registration.target;
		this.#registrations.delete(id);
		this.registrations.setState((prev) => {
			const next = new Map(prev);
			next.delete(id);
			return next;
		});
		const targetRegs = this.#targetRegistrations.get(target);
		if (targetRegs) {
			targetRegs.delete(id);
			if (targetRegs.size === 0) this.#removeListenersForTarget(target);
		}
	}
	/**
	* Ensures event listeners are attached for a specific target.
	*/
	#ensureListenersForTarget(target) {
		if (typeof document === "undefined") return;
		if (this.#targetListeners.has(target)) return;
		const keydownHandler = this.#createTargetKeyDownHandler(target);
		const keyupHandler = this.#createTargetKeyUpHandler(target);
		target.addEventListener("keydown", keydownHandler);
		target.addEventListener("keyup", keyupHandler);
		this.#targetListeners.set(target, {
			keydown: keydownHandler,
			keyup: keyupHandler
		});
	}
	/**
	* Removes event listeners for a specific target.
	*/
	#removeListenersForTarget(target) {
		if (typeof document === "undefined") return;
		const listeners = this.#targetListeners.get(target);
		if (!listeners) return;
		target.removeEventListener("keydown", listeners.keydown);
		target.removeEventListener("keyup", listeners.keyup);
		this.#targetListeners.delete(target);
		this.#targetRegistrations.delete(target);
	}
	/**
	* Creates a keydown handler for a specific target.
	*/
	#createTargetKeyDownHandler(target) {
		return (event) => {
			this.#processTargetEvent(event, target, "keydown");
		};
	}
	/**
	* Creates a keyup handler for a specific target.
	*/
	#createTargetKeyUpHandler(target) {
		return (event) => {
			this.#processTargetEvent(event, target, "keyup");
		};
	}
	/**
	* Processes keyboard events for a specific target and event type.
	*/
	#processTargetEvent(event, target, eventType) {
		const targetRegs = this.#targetRegistrations.get(target);
		if (!targetRegs) return;
		const now = Date.now();
		for (const id of targetRegs) {
			const registration = this.#registrations.get(id);
			if (!registration) continue;
			if (!isEventForTarget(event, target)) continue;
			if (!registration.options.enabled) continue;
			if (registration.options.ignoreInputs !== false) {
				if (isInputElement(event.target)) {
					if (event.target !== registration.target) continue;
				}
			}
			if (registration.options.eventType !== eventType) continue;
			const timeout = registration.options.timeout ?? DEFAULT_SEQUENCE_TIMEOUT;
			if (registration.currentIndex > 0 && now - registration.lastKeyTime > timeout) registration.currentIndex = 0;
			const expectedHotkey = registration.parsedSequence[registration.currentIndex];
			if (!expectedHotkey) continue;
			if (matchesKeyboardEvent(event, expectedHotkey, registration.options.platform)) {
				registration.lastKeyTime = now;
				registration.currentIndex++;
				if (registration.currentIndex >= registration.parsedSequence.length) {
					if (registration.options.preventDefault) event.preventDefault();
					if (registration.options.stopPropagation) event.stopPropagation();
					const context = {
						hotkey: registration.sequence.join(" "),
						parsedHotkey: registration.parsedSequence[registration.parsedSequence.length - 1]
					};
					registration.callback(event, context);
					registration.currentIndex = 0;
				}
			} else if (registration.currentIndex > 0) {
				const firstHotkey = registration.parsedSequence[0];
				if (matchesKeyboardEvent(event, firstHotkey, registration.options.platform)) {
					registration.currentIndex = 1;
					registration.lastKeyTime = now;
				} else registration.currentIndex = 0;
			}
		}
	}
	/**
	* Finds an existing registration with the same sequence and target.
	*/
	#findConflictingSequence(sequence, target) {
		const key = sequenceKey(sequence);
		for (const registration of this.#registrations.values()) if (sequenceKey(registration.sequence) === key && registration.target === target) return registration;
		return null;
	}
	/**
	* Resets all sequence progress.
	*/
	resetAll() {
		for (const registration of this.#registrations.values()) {
			registration.currentIndex = 0;
			registration.lastKeyTime = 0;
		}
	}
	/**
	* Triggers a sequence's callback programmatically from devtools.
	* Creates a synthetic KeyboardEvent from the last key in the sequence.
	*
	* @param id - The registration ID to trigger
	* @returns True if the registration was found and triggered
	*/
	triggerSequence(id) {
		const registration = this.#registrations.get(id);
		if (!registration) return false;
		const lastParsed = registration.parsedSequence[registration.parsedSequence.length - 1];
		if (!lastParsed) return false;
		const syntheticEvent = new KeyboardEvent(registration.options.eventType ?? "keydown", {
			key: lastParsed.key,
			ctrlKey: lastParsed.ctrl,
			shiftKey: lastParsed.shift,
			altKey: lastParsed.alt,
			metaKey: lastParsed.meta,
			bubbles: true,
			cancelable: true
		});
		registration.triggerCount++;
		this.registrations.setState((prev) => new Map(prev).set(id, toRegistrationView(registration)));
		const context = {
			hotkey: registration.sequence.join(" "),
			parsedHotkey: lastParsed
		};
		registration.callback(syntheticEvent, context);
		return true;
	}
	/**
	* Gets the number of registered sequences.
	*/
	getRegistrationCount() {
		return this.#registrations.size;
	}
	/**
	* Destroys the manager and removes all listeners.
	*/
	destroy() {
		for (const target of this.#targetListeners.keys()) this.#removeListenersForTarget(target);
		this.#registrations.clear();
		this.registrations.setState(() => /* @__PURE__ */ new Map());
	}
};
/**
* Gets the singleton SequenceManager instance.
* Convenience function for accessing the manager.
*/
function getSequenceManager() {
	return SequenceManager.getInstance();
}
/**
* Creates a simple sequence matcher for one-off use.
*
* This is a low-level helper that does not support ignoreInputs, target,
* or other HotkeyOptions. Callers must handle input filtering themselves
* if attaching to document.
*
* @param sequence - The sequence of hotkeys to match
* @param options - Options including timeout
* @returns An object with match() and reset() methods
*
* @example
* ```ts
* const matcher = createSequenceMatcher(['G', 'G'], { timeout: 500 })
*
* document.addEventListener('keydown', (event) => {
*   if (matcher.match(event)) {
*     console.log('Sequence matched!')
*   }
* })
* ```
*/
function createSequenceMatcher(sequence, options = {}) {
	const platform = options.platform ?? detectPlatform();
	const timeout = options.timeout ?? DEFAULT_SEQUENCE_TIMEOUT;
	const parsedSequence = sequence.map((hotkey) => parseHotkey(hotkey, platform));
	let currentIndex = 0;
	let lastKeyTime = 0;
	return {
		match(event) {
			const now = Date.now();
			if (currentIndex > 0 && now - lastKeyTime > timeout) currentIndex = 0;
			const expected = parsedSequence[currentIndex];
			if (!expected) return false;
			if (matchesKeyboardEvent(event, expected, platform)) {
				lastKeyTime = now;
				currentIndex++;
				if (currentIndex >= parsedSequence.length) {
					currentIndex = 0;
					return true;
				}
			} else if (currentIndex > 0) if (matchesKeyboardEvent(event, parsedSequence[0], platform)) {
				currentIndex = 1;
				lastKeyTime = now;
			} else currentIndex = 0;
			return false;
		},
		reset() {
			currentIndex = 0;
			lastKeyTime = 0;
		},
		getProgress() {
			return currentIndex;
		}
	};
}

//#endregion
export { SequenceManager, createSequenceMatcher, getSequenceManager };
